### ISSUE TITLE
Build fails when minify: true is set in vite.config.ts

### ISSUE THREAD

**Issue by miguelrk** [marshallswain/feathers-pinia#47]
## Problem
Enabling `'build.minify': true` in `vite.config.ts` leads to the following build error:

```error
Uncaught (in promise) TypeError: n.api.User is not a constructor
    at setupInstance (index.7faf713a.js:formatted:42194)
    at new mo (index.7faf713a.js:formatted:14221)
    at new _1e (index.7faf713a.js:formatted:42192)
    at index.7faf713a.js:formatted:13772
    at Array.forEach (<anonymous>)
    at Object.addOrUpdate (index.7faf713a.js:formatted:13767)
    at Object.<anonymous> (index.7faf713a.js:formatted:6922)
    at Object.handleFindResponse (index.7faf713a.js:formatted:13678)
    at Object.<anonymous> (index.7faf713a.js:formatted:6922)
    at index.7faf713a.js:formatted:13665
```

## Context

This is due to the `setupInstance()` method being minified to the following:

```js
// at setupInstance (index.7faf713a.js:formatted:42194)

class _1e extends Ar {
    static setupInstance(t, {store: r, models: n}) {
        return t.user && (t.user = new n.api.User(t.user).addToStore()),
        t.workspace && (t.workspace = new n.api.Workspace(t.workspace).addToStore()),
        t
    }
}
```

The following is the `vite.config.ts` build setting that causes this issue. Note that setting `build.minify: false` solves this, but obviously does not minify the code (keeps it as is, therefore no minifying error).

```ts
// vite.config.ts

const config = defineConfig({
  ...
  build: {
    minify: true, // FIXME: this breaks on build due to setupInstance
  },
})
```

Might be an easy fix, not sure what might be leading to this though.

**Comment by marshallswain** [comment]
@miguelrk what does `setupInstance` look like before compilation?  It might be an issue with class names getting rewritten in the minification process.

**Comment by miguelrk** [comment]
The following is the uncompiled source:

```typescript
class WorkspaceUser extends BaseModel {
  static setupInstance(data, { store, models }) {
    if (data.user) data.user = new models.api.User(data.user).addToStore()
    if (data?.workspace) data.workspace = new models.api.Workspace(data.workspace).addToStore()
    return data
  }
}
```

I have also tried minifying at the network level (e.g. on Cloudflare Pages Settings) rather than using vite config's `build.minify: true` and ran into the same issue. It appears that when minifying, the keys of `models.api` will be shortened by the minifier, causing issues when accessing the `User` Model. It fails in this store but I believe it will for all others as well.

The error is `Uncaught (in promise) TypeError: n.api.User is not a constructor` so specifically when renaming `models` to `n`. It's a pretty strange issue I guess @marshallswain.

**Comment by marshallswain** [comment]
Could you try adding a modelName static attribute to your User mode. 
modelName = ‘User’

I believe that I created a convention that it will use the manually
supplied modelName attribute if it exists, which gets around the
minification of constructor names.

On May 30, 2022, GitHub ***@***.***> wrote:
> The following is the uncompiled source:
>
> class WorkspaceUser extends BaseModel { static setupInstance(data, {
> store, models }) { if (data.user) data.user = new
> models.api.User(data.user).addToStore() if (data?.workspace)
> data.workspace = new models.api.Workspace(data.workspace).addToStore()
> return data } }
> I have also tried minifying at the network level (e.g. on Cloudflare
> Pages Settings) rather than using vite config's build.minify: true and
> ran into the same issue. It appears that when minifying, the keys of
> models.api will be shortened by the minifier, causing issues when
> accessing the User Model. It fails in this store but I believe it will
> for all others as well.
>
> The error is Uncaught (in promise) TypeError: n.api.User is not a
> constructor so specifically when renaming models to n. It's a pretty
> strange issue I guess @marshallswain
> <https://github.com/marshallswain>.
>

—
Reply to this email directly, view it on GitHub
<https://github.com/marshallswain/feathers-pinia/issues/47#issuecomment-
1140843150>, or unsubscribe
<https://github.com/notifications/unsubscribe-
auth/AAA7OWLM4NB5NJ3GHCIUES3VMR2B5ANCNFSM5VAH52PA>.
You are receiving this because you were mentioned.Message ID:
***@***.***>

**Comment by miguelrk** [comment]
I can confirm adding a `modelName` static attribute to **each** Model Class fixes this so based on the above: 

```typescript
class WorkspaceUser extends BaseModel {
  static modelName = 'WorkspaceUser' // ADDED THIS LINE
  static setupInstance(data, { store, models }) {
    if (data.user) data.user = new models.api.User(data.user).addToStore()
    if (data?.workspace) data.workspace = new models.api.Workspace(data.workspace).addToStore()
    return data
  }
}
```

As always @marshallswain, thank you very much. I appreciate it 🙏🏼

**Comment by marshallswain** [comment]
Awesome.  I remember how aggravating it was to encounter that the first
time.  I'm glad I had a fix ready.  :) 

On May 30, 2022, GitHub ***@***.***> wrote:
> Closed #47 <https://github.com/marshallswain/feathers-pinia/issues/47>
> as completed.
>

—
Reply to this email directly, view it on GitHub
<https://github.com/marshallswain/feathers-pinia/issues/47#event-
6704750881>, or unsubscribe
<https://github.com/notifications/unsubscribe-
auth/AAA7OWOZU6GT37MVLDBM5UDVMUDS7ANCNFSM5VAH52PA>.
You are receiving this because you were mentioned.Message ID:
<marshallswain/feathers-
***@***.***>
